app/javascript/packs/followButtons.js

```diff
@@ -0,0 +1,346 @@
+ * by obtaining the 'follow status' of each user and updating the associated buttons' UI.
+ */
+function initializeAllUserFollowButtons() {
+  const buttons = document.querySelectorAll(
+    '.follow-action-button.follow-user:not([data-fetched])',
+  );
+
+  if (buttons.length === 0) {
+    return;
+  }
+
+  const userIds = {};
+
+  Array.from(buttons, (button) => {
+    button.dataset.fetched = 'fetched';
+    const userStatus = document.body.getAttribute('data-user-status');
+
+    if (userStatus === 'logged-out') {
+      const { style } = JSON.parse(button.dataset.info);
+      addButtonFollowText(button, style);
+    } else {
+      const { id: userId } = JSON.parse(button.dataset.info);
+      if (userIds[userId]) {
+        userIds[userId].push(button);
+      } else {
+        userIds[userId] = [button];
+      }
+    }
+  });
+
+  if (Object.keys(userIds).length > 0) {
+    fetchUserFollowStatuses(userIds);
+  }
+}
+
+/**
+ * Individually fetches the current status of a follow button and updates the UI to match
+ *
+ * @param {HTMLElement} button
+ * @param {Object} buttonInfo The parsed buttonInfo object obtained from the button's data-attribute
+ */
+function fetchFollowButtonStatus(button, buttonInfo) {
+  button.dataset.fetched = 'fetched';
+
+  fetch(`/follows/${buttonInfo.id}?followable_type=${buttonInfo.className}`, {
+    method: 'GET',
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  })
+    .then((response) => response.text())
+    .then((followStatus) => {
+      updateInitialButtonUI(followStatus, button);
+      button.dataset.buttonInitialized = true;
+    });
+}
+
+/**
+ * Makes sure the initial state of follow buttons is fetched and presented in the UI.
+ * User follow buttons are initialized separately via bulk request
+ */
+function initializeNonUserFollowButtons() {
+  const nonUserFollowButtons = document.querySelectorAll(
+    '.follow-action-button:not(.follow-user)',
+  );
+
+  const user = userData();
+  nonUserFollowButtons.forEach((button) => {
+    const { info, fetched } = button.dataset;
+    const buttonInfo = JSON.parse(info);
+    if (
+      buttonInfo.className === 'Tag' &&
+      user &&
+      button.dataset.buttonInitialized !== 'true'
+    ) {
+      const initialButtonFollowState = doesUserFollowTag(user, buttonInfo.id)
+        ? 'true'
+        : 'false';
+      updateInitialButtonUI(initialButtonFollowState, button);
+      button.dataset.buttonInitialized = true;
+    } else if (fetched !== 'fetched') {
+      fetchFollowButtonStatus(button, buttonInfo);
+    }
+  });
+}
+
+initializeAllUserFollowButtons();
+initializeNonUserFollowButtons();
+listenForFollowButtonClicks();
+
+// Some follow buttons are added to the DOM dynamically, e.g. search results,
+// So we listen for any new additions to be fetched
+const observer = new MutationObserver((mutationsList) => {
```

Comment:
Refactor (potentially blocking): To avoid missing pages or future pages that might have follow buttons, we could add the `observer` to `packs/base.js` instead since that is loaded on every page, and all the code in `followButtons.js` could be imported from a non-pack ES module.

Reply:
That's a good idea 💡 

I guess the only downside is we'd be filtering through _every_ mutation in the app instead of just on these specific pages 🤔  What do you think? I agree it's a bit of a worry that we might miss a page!

Reply:
Yeah, it's a good point about filtering. But currently on the pages where it's explicitly run, it assumes there will be buttons correct?